src/ray/raylet/node_manager.cc

```diff
@@ -1176,6 +1183,52 @@ void NodeManager::ProcessSubmitTaskMessage(const uint8_t *message_data) {
   SubmitTask(Task(task_message), Lineage());
 }
 
+void NodeManager::ProcessRequestWorkerLeaseMessage(
+    const std::shared_ptr<LocalClientConnection> &client, const uint8_t *message_data) {
+  // Read the resource spec submitted by the client.
+  auto fbs_message = flatbuffers::GetRoot<protocol::WorkerLeaseRequest>(message_data);
+  rpc::Task task_message;
+  RAY_CHECK(task_message.mutable_task_spec()->ParseFromArray(
+      fbs_message->resource_spec()->data(), fbs_message->resource_spec()->size()));
+
+  // Override the task dispatch to call back to the client instead of executing the
+  // task directly on the worker. TODO(ekl) handle spilling case
+  Task task(task_message);
+  task.OnDispatchInstead([this, client](const std::shared_ptr<void> granted,
+                                        const std::string &address, int port) {
+    std::shared_ptr<Worker> client_worker = worker_pool_.GetRegisteredWorker(client);
+    if (client_worker == nullptr) {
+      client_worker = worker_pool_.GetRegisteredDriver(client);
+    }
+    if (client_worker == nullptr) {
+      RAY_LOG(ERROR) << "Lost worker for lease request " << client;
+    } else {
+      client_worker->WorkerLeaseGranted(address, port);
+      leased_workers_[port] = std::static_pointer_cast<Worker>(granted);
+    }
+  });
+  SubmitTask(task, Lineage());
+}
+
+void NodeManager::ProcessReturnWorkerMessage(const uint8_t *message_data) {
+  // Read the resource spec submitted by the client.
+  auto fbs_message = flatbuffers::GetRoot<protocol::ReturnWorkerRequest>(message_data);
+  auto worker_port = fbs_message->worker_port();
+  RAY_LOG(DEBUG) << "Return worker " << worker_port;
+  std::shared_ptr<Worker> worker = leased_workers_[worker_port];
+  leased_workers_.erase(worker_port);
+  RAY_CHECK(worker != nullptr);
+  RAY_CHECK(!worker->GetAssignedTaskId().IsNil());
```

Comment:
I think we could just call `HandleWorkerAvailable` here.

Reply:
Done